binder/src/main/java/io/grpc/binder/AndroidComponentAddress.java

```diff
@@ -177,13 +190,96 @@ public int hashCode() {
   public boolean equals(Object obj) {
     if (obj instanceof AndroidComponentAddress) {
       AndroidComponentAddress that = (AndroidComponentAddress) obj;
-      return bindIntent.filterEquals(that.bindIntent);
+      return bindIntent.filterEquals(that.bindIntent)
+          && Objects.equal(this.targetUser, that.targetUser);
     }
     return false;
   }
 
   @Override
   public String toString() {
-    return "AndroidComponentAddress[" + bindIntent + "]";
+    StringBuilder builder = new StringBuilder("AndroidComponentAddress[");
+    if (targetUser != null) {
+      builder.append(targetUser);
+      builder.append("@");
+    }
+    builder.append(bindIntent);
+    builder.append("]");
+    return builder.toString();
+  }
+
+  /**
+   * Identifies the Android user in which the bind Intent will be evaluated.
+   *
+   * <p>Returns the {@link UserHandle}, or null which means that the Android user hosting the
+   * current process will be used.
+   */
+  @Nullable
+  public UserHandle getTargetUser() {
+    return targetUser;
+  }
+
+  public static Builder newBuilder() {
+    return new Builder();
+  }
+
+  /**
+   * Fluently builds instances of {@link AndroidComponentAddress}.
+   */
+  public static class Builder {
+    Intent bindIntent = new Intent(ApiConstants.ACTION_BIND);
+    UserHandle targetUser;
+
+    /**
+     * See {@link Intent#setPackage(String)}.
+     */
+    public Builder setPackage(String pkg) {
```

Comment:
Thanks for looking!

We need a Builder because UserHandle is a new independent dimension and I don't want to double our existing 5 static factory methods. But I still want the Builder to be able to do everything they can, which means the ability to set the package and/or ComponentName without having to even know what an Intent is. At the same time many users *will* want to use the other Intent "filter" fields with the new Builder API. A setIntent(Intent) builder method would be all powerful but I don't know how to reconcile its semantics with setPackage() and setComponent()? Callers would be surprised if those fields were clobbered by calling Builder.setIntent(). forLocalComponent(), forRemoteComponent() and forContext() aren't Builder methods for a similar reason: they all act upon the same Intent fields but they have different names, which isn't the way Builders usually work.

Duplicating these few Intent setters is kinda icky but I see it as a way to prevent users from making mistakes. For example, it's always wrong to put Extras in the binding Intent because Android will silently drop them. By hiding the raw Intent inside the Builder, though, this can't happen because the putExtra()s methods aren't exposed. Intent flags are another example: Almost all the Intent#FLAG_XXX flags don't make sense for bindService() Intents and this way I can expose setters for just the ones that do (not present in this PR just because nobody as asked for it yet).

LMK if you buy all this.

Reply:
> I don't want to double our existing 5 static factory methods

That explains it.

> A setIntent(Intent) builder method would be all powerful but I don't know how to reconcile its semantics with setPackage() and setComponent()?

This seems to be a problem of your own making. If you duplicated the existing static factory methods or a more limited `setIntentFromComponent(ComponentName)`+`setIntent(Intent)` then there's no problem. Every set() completely replaces the Intent.

> Duplicating these few Intent setters is kinda icky but I see it as a way to prevent users from making mistakes.

It seems to expose users to a lot more than the existing methods. The earlier way guided you and had you ignore most of the Intent, unless you knew enough to make an Intent. The new way seems to assume you know enough to make an Intent, but then assumes you don't know how to make an Intent.

Today the user can either: specify the Component (in various ways; seems the common cases) or provide the entire Intent. Now you're exposing lots of other ways to tweak it and I don't see how it is better than the already-available `setBindIntent()`. It's like we're encouraging people to make a really custom Intent instead of just setting a ComponentName. Unless we're saying we want to deprecate `setBindIntent()`, I don't see what we're gaining here.

Reply:
Naming convenience builder methods like setIntentFromXXX() is a nice way to solve the problem. Thanks!